Add SHORT_QUERY_ADJUST support in Bio.Blast.NCBIWWW

MarkusPiotrowski opened the pull request:
This pull request addresses issue #2049. It allows to use the SHORT_QUERY_ADJUST parameter of BLAST's common URL API in `Bio.Blast.NCBIWWW`via a new keyword parameter `short_query`. 
Unfortunately, SHORT_QUERY_ADJUST is faulty for blastn, thus a workaround for blastn was implemented.

Also, `test_NCBI_qblast.py` was corrected (did fail on two tests) and updated.

This will close #2049.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file and understand that AppVeyor and
TravisCI will be used to confirm the Biopython unit tests and ``flake8`` style
checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

MarkusPiotrowski commented:
Need to implement sequence length dependency for blastn workaround.

MarkusPiotrowski commented:
Hey, how comes a D400 in my commit from a file I didn't touch? :-( 


<pre>
Tests/test_PDB_parse_pdb_header.py:53:1: D400 First line should end with a period
ERROR: InvocationError for command /home/travis/build/biopython/biopython/.tox/style/bin/flake8 Tests/ (exited with code 1)
</pre>

I'm going to correct this...

MarkusPiotrowski commented:
Now I got caught by:
```
======================================================================
ERROR: test_simple_scatter_plot_1 (test_GraphicsGeneral.ComparativeTest)
Test creation of a simple ScatterPlot with one list.
----------------------------------------------------------------------
Traceback (most recent call last):
...
ZeroDivisionError: float division by zero
```

Sigh...

peterjc commented:
It seems #2040 is causing a disproportionate amount of pain, we should fix it ASAP.

peterjc commented:
I would not include the workaround.

Have you reported the blastn issue to the BLAST team? They can likely fix this on their server (relatively quickly), which seem far better in the long run that having a workaround in (some versions of) biopython.

MarkusPiotrowski commented:
It seems that they are not interested in fixing this issue...
Yes, I have contacted them and also give them the likely reason why it fails: It seems that they are just using the blastp SHORT_QUERY_ADJUST values in blastn (where the word size is to small for blastn and thus throws an error).

> Dear Sir or Madame,
> 
> I want to report an error with the common URL API.
> 
> The parameter SHORT_QUERY_ADJUST does not work for blastn, obviously the parameter adjustment procedure uses the parameters for blastp, which results in a blastn error (WORD_SIZE must be >=4).
> 
> A working blastp search with SHORT_QUERY_ADJUST:
> 
> https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastp&DATABASE=nr&QUERY=ICWENK&HITLIST_SIZE=5&SHORT_QUERY_ADJUST=on&CMD=Put
> 
> Now a failing blastn search with SHORT_QUERY_ADJUST:
> 
> https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&DATABASE=nt&QUERY=ATGCTGATCGGATCAACG&HITLIST_SIZE=5&SHORT_QUERY_ADJUST=on&CMD=Put
> 
> Message ID#36 Error: Cannot validate the Blast options: Word-size must be 4 or greater for nucleotide comparison
> 
> Taking a look at the html source code on the Format Request page, you see that the parameters were wrongly set for blastn:
>  
> 
> \<input name="EXPECT" type="hidden" value="200000"/>
> ...
> \<input name="WORD_SIZE" type="hidden" value="2"/>
> ...
> \<input name="MATCH_SCORES" type="hidden" value=""/>
> 
> 
> EXPECT should be 1000, WORD_SIZE 7, and MATCH_SCORES 1, -3. There may be other parameters which require changes (e.g. filtering and masking.)
> 
> Sincerely yours,
> Markus Piotrowski

The reply was that they are not officially supporting this 'SHORT_QUERY_ADJUST' and they recommend to set the respective values manually (what I did in the workaround).

> Hi Markus,
> 
> The fact is, we only officially support the parameters outlined here,
> 
> https://ncbi.github.io/blast-cloud/dev/api.html
> 
> For short queries, I suggest you explicitly set EXPECT, WORD_SIZE, etc.
> Best regards,
> Wayne

What is the problem with the workaround? If it is ommitted people will not be able to use `SHORT_QUERY_ADJUST` with `blastn`.

peterjc commented:
> What is the problem with the workaround? If it is ommitted people will not be able to use SHORT_QUERY_ADJUST with blastn.

Exactly 👍 

It sounds like the NCBI would prefer we didn't expose the ``SHORT_QUERY_ADJUST`` parameter at all. The problem is once we do add it to Biopython, people will expect us to support it *and the workaround*.

If you do want to add the undocumented parameter, I would say explicitly in our documentation that at the time of writing it is not supported by the NCBI, and furthermore, note  it does not (currently) work with BLASTN, and document the workaround they suggest.

But perhaps I'm being too cautious?-By all means ask another Biopython developer or two for their perspective.

MarkusPiotrowski commented:
I'm sorry, Peter, I didn't thought that adding another 'undocumented' parameter would harm, since the 'old' NCBI's [qblast API](https://web.archive.org/web/20070701014147/http://www.ncbi.nlm.nih.gov/BLAST/Doc/node5.html) (on which `Bio.Blast.NCBIWWW` is based) did support more or other parameters than are supported in the 'new' [common URL API](https://ncbi.github.io/blast-cloud/dev/api.html). 

So we may think about a new module which is fully compliant with NCBI's Common URL API.

peterjc commented:
That's a good counter argument. OK, stick a cautionary note in the docstring, and go ahead with the code as is.

MarkusPiotrowski commented:
@bow I fear I don't understand what you have in mind with the toggle-able switch.
Let me explain what I did and why:
1. I added the possibility to use the parameter `SHORT_QUERY_ADJUST`. 
`SHORT_QUERY_ADJUST` is an (undocumented) parameter for NCBI's remote web searches. It works nicely for **blastp, tblastn and blastx** and it (automatically) relaxes some search parameters like word size and e-value if the sequence is shorter than 31 characters.
2. However, NCBI has made an error in the implementation for **blastn**. They use the wrong parameters and thus a BLAST error is invoked (word-size to small). To bypass this problem, I set the respective parameters manually, but only for this case, so that users of `Bio.Blast.NCBIWWW` didn't get caught by the problem. But we don't set the parameters manually in case of **blastp** etc.

`SHORT_QUERY_ADJUST` is off by default when doing remote web searches.

MarkusPiotrowski commented:
@bow BTW the cautionary note that @peterjc is requesting is not this `BiopythonWarning` but about using undocumented parameters of NCBIs Common URL API. So I added the following paragraph:

> Please note that NCBI uses the new Common URL API for BLAST searches
> on the internet (http://ncbi.github.io/blast-cloud/dev/api.html). Thus,
>  some of the parameters used by this function are not or not longer
>  officially supported by NCBI. Although they are still functioning, this
>  may change in the future.

However, I corrected the `BiopythonWarning` message as you have requested.

peterjc commented on Bio/Blast/NCBIWWW.py:
```diff
@@ -42,13 +46,18 @@ def qblast(program, database, sequence, url_base=NCBI_BLAST_URL,
-    format_object='Alignment'. For more details, please see
+    Please note that NCBI uses the new Common URL API for BLAST searches
+    on the internet (http://ncbi.github.io/blast-cloud/dev/api.html). Thus,
+    some of the parameters used by this function are not or not longer
```
Something odd with the wording here, "are not are not longer officially supported".

Maybe just "are no longer officially supported" is close to your intent?

MarkusPiotrowski replied:
"Are not *or* not longer" ("are not/not longer supported") to say that some of them were supported in the past (old qblast) but some of them (e.g. SHORT_QUERY_ADJUST) have never been officially supported (or documented)

peterjc replied:
Sorry, I missed the "or" in the earlier quote.

How about "... are not (or are no longer) officially supported." here?

MarkusPiotrowski replied:
OK.

peterjc commented on Bio/Blast/NCBIWWW.py:
```diff
@@ -80,6 +91,23 @@ def qblast(program, database, sequence, url_base=NCBI_BLAST_URL,
+            nucl_reward = 1
+            filter = None
+            lcase_mask = None
+            warnings.warn('"short_query" for blastn is wronlgy implemented by '
```
Maybe "is not currently implemented by the NCBI"?

Also, is the rest of this warning clear enough that the workaround will ignore some values the user might have set (expect, word size, etc)?

MarkusPiotrowski replied:
I think that is the same what is happening when 'SHORT_QUERY_ADJUST' is used in **blastp* etc. I can put this in the docstring if you like?

MarkusPiotrowski replied:
Actually, it is implemented, but wrongly. If you use SHORT_QUERY_ADJUST with **blastn** you get a BLAST error. 

This is the warning as have it now:

> '"SHORT_QUERY_ADJUST" is incorrectly implemented '
>                           '(by NCBI) for blastn. We bypass the problem by '
>                           'manually adjusting the search parameters. Thus, '
>                           'results may slightly differ from web page '
>                           'searches.'

peterjc commented on Tests/test_NCBI_qblast.py:
```diff
@@ -1,7 +1,9 @@
+# This file is part of the Biopython distribution and governed by your
+# choice of the "Biopython License Agreement" or the "BSD 3-Clause License".
+# Please see the LICENSE file that should have been included as part of this
+# package.
```
Can we do the dual licensing separately, with a more explicit audit trail?

MarkusPiotrowski replied:
OK, it thought because you are the author we can do this in one go.

peterjc replied:
The main author yes, but there are half a dozen or so contributors to this file according to GitHub, so I'd prefer to do it a little more formally. I've got this fairly automated now ;)

MarkusPiotrowski commented:
I implemented the requested and suggested changes (mostly...).

peterjc commented:
Looks good, over to you @bow

bow commented:
@MarkusPiotrowski Aha, thanks for the explanation. I did miss some parts of it; now I see why having the parameter I suggested earlier does not really make sense.

Anyway, it looks ok for merging now. I'll do a rebase very soon :).
